### PR TITLE
wallet: bugfix, disallow migration of invalid scripts

### PR DESCRIPTION
**Pull request by furszy**
Fixing #28057.

The legacy wallet allows to import any raw script (#28126), without
checking if it was valid or not. Appending it to the watch-only set.

This causes a crash in the migration process because we are only
expecting to find valid scripts inside the legacy spkm.

These stored scripts internally map to `ISMINE_NO` (same as if they
weren't stored at all..).

So we need to check for these special case, and take into account that
the legacy spkm could be storing invalid not watched scripts.

Which, in code words, means `IsMineInner()` returning
`IsMineResult::INVALID` for them.

Note:
To verify this, can run the test commit on top of master.
`wallet_migration.py` will crash without the bugfix commit.

### CODE REVIEW

**Review by bitcoinfinancier — APPROVED** [review]

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: test/functional/wallet_migration.py [hunk 1]
```diff
@@ -674,6 +674,12 @@ def send_to_script(script, amount):
         wallet.rpc.importaddress(address=script_wsh_pkh.hex(), label="raw_spk2", rescan=True, p2sh=False)
         assert_equal(wallet.getbalances()['watchonly']['trusted'], 5)
 
+        # Import sh(pkh()) script with the p2sh flag enabled.
+        # This will wrap the script under another sh level, which is invalid!, and store it inside the wallet.
+        # The migration process must skip the invalid scripts and the addressbook records linked to them.
+        # They are not being watched by the current wallet, nor should be watched by the migrated one.
+        wallet.rpc.importaddress(address=script_sh_pkh.hex(), label="boom_script", rescan=True, p2sh=True)
```
[the remark this answers is not in the document]
Comment (furszy) [reply test/functional/wallet_migration.py]: yep. Pushed.

**Comment by achow101** [comment]
ACK 26f91a56afd01ce11245944c66361da9aaa6a71e

File: test/functional/wallet_migration.py [hunk 2]
```diff
@@ -674,6 +674,12 @@ def send_to_script(script, amount):
         wallet.rpc.importaddress(address=script_wsh_pkh.hex(), label="raw_spk2", rescan=True, p2sh=False)
         assert_equal(wallet.getbalances()['watchonly']['trusted'], 5)
 
+        # Import sh(pkh()) script with the p2sh flag enabled.
+        # This will wrap the script under another sh level, which is invalid!, and store it inside the wallet.
+        # The migration process must skip the invalid scripts and the addressbook records linked to them.
+        # They are not being watched by the current wallet, nor should be watched by the migrated one.
+        wallet.rpc.importaddress(address=script_sh_pkh.hex(), label="boom_script", rescan=False, p2sh=True)
```
Comment (achow101) [inline test/functional/wallet_migration.py:681]: Perhaps check that this label doesn't appear in any of the migrated wallets?

**Review by furszy — COMMENTED** [review]
Feedback tackled, thanks achow101.

Expanded test coverage with the requested label check and few more validations. Now the test:
1) Verifies that neither the invalid double sh script label, nor the `addr(<script_hash_addr>)` descriptor, are contained by any of the migrated wallets.
2) Verifies that the valid sh script (the original single sh script that is imported at the same time that the invalid one) and its address book record are contained by the migrated watch-only wallet.

Also, while was doing this. Corrected another behavior; the invalid script label was being retained in the main wallet (not on the watch-only one), which was not intended.

**Comment by fanquake** [comment]
@MarcoFalke does this fix #28057 in it's current state?

File: test/functional/wallet_migration.py [hunk 3]
```diff
@@ -674,6 +675,21 @@ def send_to_script(script, amount):
...
+        # The migration process must skip the invalid scripts and the addressbook records linked to them.
+        # They are not being watched by the current wallet, nor should be watched by the migrated one.
+        label_sh_pkh = "raw_sh_pkh"
+        script_pkh = key_to_p2pkh_script(df_wallet.getaddressinfo(df_wallet.getnewaddress())["pubkey"])
+        script_sh_pkh = script_to_p2sh_script(script_pkh)
+        addy_script_sh_pkh = script_to_p2sh(script_pkh)  # valid script address
+        addy_script_double_sh_pkh = script_to_p2sh(script_sh_pkh)  # invalid script address
+
+        # Note: 'importaddress()' will add two scripts, a valid one sh(pkh()) and an invalid one 'sh(sh(pkh()))'.
+        #       Both of them will be stored with the same addressbook label. And only the former one should
```
Comment (achow101) [inline test/functional/wallet_migration.py:689]: In 21ac6b2932b0f5ea9da4c251c65f4b88dfe8e983 "test: wallet, verify migration doesn't crash for an invalid script"

s/former/latter

**Review by furszy — COMMENTED** [review]
Updated per feedback. Thanks achow.

File: test/functional/wallet_migration.py (hunk 3 above)
Comment (furszy) [reply test/functional/wallet_migration.py]: done

**Comment by achow101** [comment]
ACK 8e7e3e614955e60d3bf9e9a481ef8916bf9e22d9

**Review by bitcoinfinancier — APPROVED** [review]

**Review by jonesk7734 — COMMENTED** [review]
Ok

**Comment by fanquake** [comment]
Backported to 25.x in #28487.

**Pull request merged**